File: internal/plugins/jest/rules/no_test_prefixes/no_test_prefixes_test.go
package no_test_prefixes_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/no_test_prefixes"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoTestPrefixesRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_test_prefixes.NoTestPrefixesRule,
		[]rule_tester.ValidTestCase{
			{Code: `describe("foo", function () {})`},
			{Code: `it("foo", function () {})`},
			{Code: `it.concurrent("foo", function () {})`},
			{Code: `test("foo", function () {})`},
			{Code: `test.concurrent("foo", function () {})`},
			{Code: `describe.only("foo", function () {})`},
			{Code: `it.only("foo", function () {})`},
			{Code: `it.each()("foo", function () {})`},
			{Code: "it.each``(\"foo\", function () {})"},
			{Code: `test.only("foo", function () {})`},
			{Code: `test.each()("foo", function () {})`},
			{Code: "test.each``(\"foo\", function () {})"},
			{Code: `describe.skip("foo", function () {})`},
			{Code: `it.skip("foo", function () {})`},
			{Code: `test.skip("foo", function () {})`},
			{Code: `foo()`},
			{Code: `[1,2,3].forEach()`},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code:   `fdescribe("foo", function () {})`,
				Output: []string{`describe.only("foo", function () {})`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 1, Column: 1},
				},
			},
			{
				Code:   `xdescribe.each([])("foo", function () {})`,
				Output: []string{`describe.skip.each([])("foo", function () {})`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 1, Column: 1},
				},
			},
			{
				Code:   `fit("foo", function () {})`,
				Output: []string{`it.only("foo", function () {})`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 1, Column: 1},
				},
			},
			{
				Code:   `xdescribe("foo", function () {})`,
				Output: []string{`describe.skip("foo", function () {})`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 1, Column: 1},
				},
			},
			{
				Code:   `xit("foo", function () {})`,
				Output: []string{`it.skip("foo", function () {})`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 1, Column: 1},
				},
			},
			{
				Code:   `xtest("foo", function () {})`,
				Output: []string{`test.skip("foo", function () {})`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 1, Column: 1},
				},
			},
			{
				Code:   "xit.each``(\"foo\", function () {})",
				Output: []string{"it.skip.each``(\"foo\", function () {})"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 1, Column: 1},
				},
			},
			{
				Code:   "xtest.each``(\"foo\", function () {})",
				Output: []string{"test.skip.each``(\"foo\", function () {})"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 1, Column: 1},
				},
			},
			{
				Code:   `xit.each([])("foo", function () {})`,
				Output: []string{`it.skip.each([])("foo", function () {})`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 1, Column: 1},
				},
			},
			{
				Code:   `xtest.each([])("foo", function () {})`,
				Output: []string{`test.skip.each([])("foo", function () {})`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 1, Column: 1},
				},
			},
			{
				Code: `
import { xit } from '@jest/globals';

xit("foo", function () {})
`,
				Output: []string{`
import { xit } from '@jest/globals';

it.skip("foo", function () {})
`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 4, Column: 1},
				},
			},
			{
				Code: `
import { xit as skipThis } from '@jest/globals';

skipThis("foo", function () {})
`,
				Output: []string{`
import { xit as skipThis } from '@jest/globals';

it.skip("foo", function () {})
`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 4, Column: 1},
				},
			},
			{
				Code: `
import { fit as onlyThis } from '@jest/globals';

onlyThis("foo", function () {})
`,
				Output: []string{`
import { fit as onlyThis } from '@jest/globals';

it.only("foo", function () {})
`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "usePreferredName", Line: 4, Column: 1},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/no-test-prefixes.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-test-prefixes', {} as never, {
  valid: [
    { code: 'describe("foo", function () {})' },
    { code: 'it("foo", function () {})' },
    { code: 'it.concurrent("foo", function () {})' },
    { code: 'test("foo", function () {})' },
    { code: 'test.concurrent("foo", function () {})' },
    { code: 'describe.only("foo", function () {})' },
    { code: 'it.only("foo", function () {})' },
    { code: 'it.each()("foo", function () {})' },
    { code: 'it.each``("foo", function () {})' },
    { code: 'test.only("foo", function () {})' },
    { code: 'test.each()("foo", function () {})' },
    { code: 'test.each``("foo", function () {})' },
    { code: 'describe.skip("foo", function () {})' },
    { code: 'it.skip("foo", function () {})' },
    { code: 'test.skip("foo", function () {})' },
    { code: 'foo()' },
    { code: '[1,2,3].forEach()' },
  ],
  invalid: [
    {
      code: 'fdescribe("foo", function () {})',
      output: 'describe.only("foo", function () {})',
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'describe.only' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'xdescribe.each([])("foo", function () {})',
      output: 'describe.skip.each([])("foo", function () {})',
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'describe.skip.each' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'fit("foo", function () {})',
      output: 'it.only("foo", function () {})',
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'it.only' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'xdescribe("foo", function () {})',
      output: 'describe.skip("foo", function () {})',
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'describe.skip' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'xit("foo", function () {})',
      output: 'it.skip("foo", function () {})',
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'it.skip' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'xtest("foo", function () {})',
      output: 'test.skip("foo", function () {})',
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'test.skip' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'xit.each``("foo", function () {})',
      output: 'it.skip.each``("foo", function () {})',
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'it.skip.each' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'xtest.each``("foo", function () {})',
      output: 'test.skip.each``("foo", function () {})',
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'test.skip.each' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'xit.each([])("foo", function () {})',
      output: 'it.skip.each([])("foo", function () {})',
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'it.skip.each' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: 'xtest.each([])("foo", function () {})',
      output: 'test.skip.each([])("foo", function () {})',
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'test.skip.each' },
          column: 1,
          line: 1,
        },
      ],
    },
    {
      code: `
        import { xit } from '@jest/globals';

        xit("foo", function () {})
      `,
      output: `
        import { xit } from '@jest/globals';

        it.skip("foo", function () {})
      `,
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'it.skip' },
          column: 1,
          line: 3,
        },
      ],
    },
    {
      code: `
        import { xit as skipThis } from '@jest/globals';

        skipThis("foo", function () {})
      `,
      output: `
        import { xit as skipThis } from '@jest/globals';

        it.skip("foo", function () {})
      `,
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'it.skip' },
          column: 1,
          line: 3,
        },
      ],
    },
    {
      code: `
        import { fit as onlyThis } from '@jest/globals';

        onlyThis("foo", function () {})
      `,
      output: `
        import { fit as onlyThis } from '@jest/globals';

        it.only("foo", function () {})
      `,
      errors: [
        {
          messageId: 'usePreferredName',
          data: { preferredNodeName: 'it.only' },
          column: 1,
          line: 3,
        },
      ],
    },
  ],
});
